content/blog/overview-package-management-python.md

```diff
@@ -0,0 +1,363 @@
+- errors are very cryptic: I made a typo while adding a dependency in `Pipfile` resulting in an invalid TOML variable and all I got was
+two tracebacks for a total of 70 lines with the reason on the last line. How about only showing me the reason instead?
+- `pipenv update` doesn't tell what was actually updated...?
+- `pipenv update django` updates all the packages instead of only django....?
+
+## poetry
+I believe I first heard of [poetry](https://github.com/sdispater/poetry) while looking up [Black](https://github.com/ambv/black) and noticing
+a `pyproject.toml`.
+
+[PEP 518](https://www.python.org/dev/peps/pep-0518/) introduced `pyproject.toml`:
+
+> This PEP specifies how Python software packages should specify what build dependencies they have in order
+> to execute their chosen build system.
+> As part of this specification, a new configuration file is introduced for software packages to use to specify
+> their build dependencies (with the expectation that the same configuration file will be used for future configuration details).
+
+The interesting part of that PEP for the purpose of this article is the [tool section](https://www.python.org/dev/peps/pep-0518/#tool-table).
+
+Rather than having one config file for each tool like [Flake8](http://flake8.pycqa.org/en/latest/index.html#) or Black, having
+all of them consolidated in one file would be very valuable.
+
+Poetry uses this file to list the dependencies, keeping everything neatly in place.
+
+There are two ways to get start with poetry:
+
+- `poetry new SOME_NAME`: will create a folder named SOME_NAME with some
+- `poetry init`: an interactive way to setup your project and some dependencies
+
+The downside of the `poetry init` is that the package search is not that great: searching for `django` will
+return a list of 99 packages, all of them being `django-something` and not the actual django. I don't know what Pypi
+offers as an API so maybe it is a limitation of the API but even then, showing 99 results is not very user friendly and
+if there is an exact match it should show it to you.
+The help text says I can enter the exact package name but typing `django` tries to autocomplete with the package names
+from the list and results in an error:
+
+```bash
+ > django
+Value "django" is invalid
+
+Enter package # to add, or the complete package name if it is not listed:
+ [ 0] django-bagou
+ [ 1] django-maro
+ ... 99 packages
+```
+
+Poetry will also create virtualenv automatically when running `install`, `add` or `remove` with the current version of `python`:
+you cannot currently create a virtualenv with a different version.
+
+You are supposed to activate a shell with `poetry shell` but it breaks if you using `$WORKON_HOME` for other tools:
+https://github.com/sdispater/poetry/issues/214 so I ended up ignoring the virtualenv related commands. I didn't spend time looking
+into it so there might be an easy fix but it should work out of the box with the rest of the ecosystem.
```

Comment:
> it should work out of the box
Keep in mind that it's a 0.11.5 . As far as I know, that means "non-production ready" :)